### PR TITLE
[tests-only][full-ci]: use graph endpoint to create public link

### PR DESCRIPTION
**Pull request by ishabaral**
## Description
Replace UI with graph endpoint for creating public link

## Related Issue

- Part of https://github.com/owncloud/web/issues/10949

## How Has This Been Tested?

- test environment:
- test case 1: locally
- test case 2: CI
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests
- [ ] Documentation
- [ ] Maintenance (e.g. dependency updates or tooling)

## Open tasks:

- [ ] ...

### CODE REVIEW

File: tests/e2e/cucumber/features/app-provider/lock.feature [hunk 1]
```diff
@@ -37,7 +37,7 @@ Feature: lock
     And "Alice" should not be able to manage share of a file "test.odt" for user "Brian"
 
     # checking that sharing and creating link of the locked file is possible
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/app-provider/lock.feature:40]: should be UI step

File: tests/e2e/cucumber/features/file-action/rename.feature [hunk 2]
```diff
@@ -19,7 +19,7 @@ Feature: rename
       | resource | recipient | type | role     |
       | folder   | Brian     | user | Can edit |
     And "Alice" opens the "files" app
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/file-action/rename.feature:22]:
```suggestion
    And "Alice" creates a public link of following resource using API
    And "Alice" opens the "files" app
```

File: tests/e2e/cucumber/features/navigation/shortcut.feature [hunk 3]
```diff
@@ -21,7 +21,7 @@ Feature: Users can create shortcuts for resources and sites
       | resource       | recipient | type | role     |
       | testavatar.jpg | Brian     | user | Can view |
     And "Alice" opens the "files" app
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/navigation/shortcut.feature:24]:
```suggestion
    And "Alice" creates a public link of following resource using API
    And "Alice" opens the "files" app
```

File: tests/e2e/cucumber/features/shares/link.feature [hunk 4]
```diff
@@ -20,7 +20,7 @@ Feature: link
       | folderPublic/lorem.txt | lorem ipsum |
 
     And "Alice" opens the "files" app
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/shares/link.feature:23]: all link creation steps should be from UI in this feature
```suggestion
    And "Alice" creates a public link of following resource using API
```

File: tests/e2e/cucumber/features/shares/internalLink.feature [hunk 5]
```diff
@@ -14,7 +14,7 @@ Feature: internal link share
       | resource | recipient | type | role     |
       | myfolder | Brian     | user | Can edit |
     And "Alice" opens the "files" app
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/shares/internalLink.feature:17]: let's use UI step here
```suggestion
    And "Alice" creates a public link of following resource using the sidebar panel
```

File: tests/e2e/cucumber/features/spaces/internalLink.feature [hunk 6]
```diff
@@ -26,7 +26,7 @@ Feature: internal link share in project space
     And "Alice" navigates to the project space "marketing.1"
     # internal link to space
 
-    And "Alice" creates a public link for the space with password "%public%" using the sidebar panel
+    And "Alice" creates a public link for the space "Marketing" with password "%public%" using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/spaces/internalLink.feature:29]: let's use UI step here
```suggestion
    And "Alice" creates a public link for the space "Marketing" with password "%public%" using the sidebar panel
```

File: tests/e2e/cucumber/features/app-provider/officeSuites.feature [hunk 7]
```diff
@@ -18,7 +18,7 @@ Feature: Integrate with online office suites like Collabora and OnlyOffice
     When "Alice" creates the following resources
       | resource         | type         | content              |
       | OpenDocument.odt | OpenDocument | OpenDocument Content |
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/app-provider/officeSuites.feature:21]: let's use UI step in this feature

File: tests/e2e/cucumber/features/spaces/publicLink.feature [hunk 8]
```diff
@@ -31,7 +31,7 @@ Feature: spaces public link
       | resource       |
       | simple.pdf     |
       | testavatar.jpg |
-    And "Alice" creates a public link for the space with password "%public%" using the sidebar panel
+    And "Alice" creates a public link for the space "team" with password "%public%" using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/spaces/publicLink.feature:34]: link creation should be from UI step in this feature

File: tests/e2e/cucumber/features/user-settings/languageChange.feature [hunk 9]
```diff
@@ -39,7 +39,7 @@ Feature: language settings
       | filesForUpload/lorem.txt | lorem.txt |
 
     And "Alice" opens the "files" app
-    And "Alice" creates a public link of following resource using the sidebar panel
+    And "Alice" creates a public link of following resource using API
```
Comment (saw-jan) [inline tests/e2e/cucumber/features/user-settings/languageChange.feature:42]:
```suggestion
    And "Alice" creates a public link of following resource using API
    And "Alice" opens the "files" app
```

File: tests/e2e/cucumber/steps/ui/links.ts [hunk 10]
```diff
@@ -20,17 +20,6 @@ When(
   }
 )
 
-When(
-  '{string} creates a public link for the space with password {string} using the sidebar panel',
```
Comment (saw-jan) [inline tests/e2e/cucumber/steps/ui/links.ts:24]: we need this step

File: tests/e2e/support/objects/app-files/link/actions.ts [hunk 11]
```diff
@@ -191,6 +192,8 @@ export const changeName = async (args: changeNameArgs): Promise<string> => {
     }
     await sidebar.open({ page: page, resource: resourceName })
     await sidebar.openPanel({ page: page, name: 'sharing' })
+  } else {
+    await openSharingPanel(page)
```
Comment (saw-jan) [inline tests/e2e/support/objects/app-files/link/actions.ts:196]: why was this required?

File: tests/e2e/support/api/share/share.ts [hunk 12]
```diff
@@ -179,3 +193,91 @@ export const addMembersToTheProjectSpace = async ({
...
+  const linksEnvironment: LinksEnvironment = new LinksEnvironment()
+  linksEnvironment.createLink({
+    key: name,
+    link: { name: name, url: webUrl }
+  })
+
+  checkResponseStatus(response, 'Failed while creating public link share')
+}
+
+export const createPublicLinkForSpace = async ({
```
Comment (saw-jan) [inline tests/e2e/support/api/share/share.ts:246]:
```suggestion
export const createSpaceLinkShare = async ({
```

**Review by saw-jan — CHANGES_REQUESTED** [review]

File: tests/e2e/support/objects/app-files/link/actions.ts (hunk 11 above)
Comment (ishabaral) [reply tests/e2e/support/objects/app-files/link/actions.ts:196]: In `publicLink.feature` when using webUI to create public link for the space, the sharing panel is already opened, which is followed by rename  step in line 135. So on removing the webUI step for creating public link, the test case fails at line 135. So i need to add `await openSharingPanel(page)`
```
134.  And "Alice" creates a public link for the space "team" with password "%public%" using the sidebar panel
135. And "Alice" renames the most recently created public link of space to "spaceLink"
```

Comment (ishabaral) [reply tests/e2e/support/objects/app-files/link/actions.ts:196]: But as per your review we don't need to use API steps there in `publicLink.feature`, so may be this is not needed.

Comment (saw-jan) [reply tests/e2e/support/objects/app-files/link/actions.ts]: if not used then yes, let's not add it.

File: tests/e2e/cucumber/steps/api.ts [hunk 13]
```diff
@@ -286,3 +286,23 @@ Given(
     }
   }
 )
+
+Given(
+  '{string} creates a public link of following resource using API',
```
Comment (saw-jan) [inline tests/e2e/cucumber/steps/api.ts:291]: this should go to the `api.ts`

File: tests/e2e/cucumber/steps/ui/links.ts [hunk 14]
```diff
@@ -21,8 +21,13 @@ When(
 )
 
 When(
-  '{string} creates a public link for the space with password {string} using the sidebar panel',
-  async function (this: World, stepUser: string, password: string): Promise<void> {
+  '{string} creates a public link for the space {string} with password {string} using the sidebar panel',
```
Comment (saw-jan) [inline tests/e2e/cucumber/steps/ui/links.ts:24]: let's not change this step.
```suggestion
  '{string} creates a public link for the space with password {string} using the sidebar panel',
```

File: tests/e2e/support/api/share/share.ts [hunk 15]
```diff
@@ -35,6 +35,20 @@ export const shareRoles: Readonly<{
   'Secret File Drop': 'uploader'
 } as const
 
+export const sharingLinkType: Readonly<{
```
Comment (saw-jan) [inline tests/e2e/support/api/share/share.ts:38]:
```suggestion
export const linkShareRoles: Readonly<{
```

File: tests/e2e/support/api/share/share.ts [hunk 16]
```diff
@@ -179,3 +193,54 @@ export const addMembersToTheProjectSpace = async ({
...
   checkResponseStatus(response, 'Failed while adding members to project space')
 }
+
+export const createLinkShare = async ({
+  user,
+  path,
+  password,
+  name,
+  role,
+  spaceName
```
Comment (saw-jan) [inline tests/e2e/support/api/share/share.ts:203]: can we add default value here?
```suggestion
  role = "view",
  spaceName = "personal"
```

**Review by saw-jan — CHANGES_REQUESTED** [review]

File: tests/e2e/cucumber/steps/api.ts (hunk 13 above)
Comment (ishabaral) [reply tests/e2e/cucumber/steps/api.ts:291]: isn't this in api.ts already?

File: tests/e2e/support/api/share/share.ts (hunk 16 above)
Comment (ishabaral) [reply tests/e2e/support/api/share/share.ts:203]: i don't think we can because according to our feature file, the role is sometimes edit also (`rename.feature:21`). So role can't have default value. 
In case of spaceName, we can have a default value for now since we're creating a public link of resources made in personal space. But what if need later for 'project' space? Shall I make it default?

File: tests/e2e/cucumber/steps/api.ts (hunk 13 above)
Comment (saw-jan) [reply tests/e2e/cucumber/steps/api.ts:291]: my eyes :eyeglasses: . sorry for that :sweat_smile:

File: tests/e2e/support/api/share/share.ts (hunk 16 above)
Comment (saw-jan) [reply tests/e2e/support/api/share/share.ts:203]: yes, default values are meant to be used when no optional params are provided (here: `role` and `spaceName`). if we want different role or spaceName other than default then we should be able to provide them as parameters in function call, right?

Comment (ishabaral) [reply tests/e2e/support/api/share/share.ts]: right :)

File: tests/e2e/support/api/share/share.ts [hunk 17]
```diff
@@ -179,3 +193,56 @@ export const addMembersToTheProjectSpace = async ({
...
   checkResponseStatus(response, 'Failed while adding members to project space')
 }
+
+export const createLinkShare = async ({
+  user,
+  path,
+  password,
+  name,
+  role = 'Can view',
+  spaceType = 'personal'
```
Comment (saw-jan) [inline tests/e2e/support/api/share/share.ts:203]:
```suggestion
  spaceName = 'Personal'
```

File: tests/e2e/support/api/share/share.ts [hunk 18]
```diff
@@ -179,3 +193,56 @@ export const addMembersToTheProjectSpace = async ({
...
+}: {
+  user: User
+  path: string
+  password: string
+  name: string
+  role?: string
+  spaceType?: string
+}): Promise<void> => {
+  const spaceName = spaceType === 'personal' ? user.displayName : spaceType
```
Comment (saw-jan) [inline tests/e2e/support/api/share/share.ts:212]:
```suggestion
  let spaceType
  if (spaceName.toLowerCase() === "personal") {
    spaceName = user.displayName
    spaceType = spaceName.toLowerCase()
  } else {
    spaceType = "project"
  }
```

**Review by saw-jan — APPROVED** [review]
LGTM

**Review by ScharfViktor — APPROVED** [review]
